react-components/src/components/Reveal3DResources/useRemoveNonReferencedModels.ts

```diff
@@ -0,0 +1,92 @@
+import { Cognite3DViewer, CogniteModel, Image360Collection } from '@cognite/reveal';
+import { AddImageCollection360Options, TypedReveal3DModel } from './types';
+import { is360ImageDataModelAddOptions, is3dModelOptions } from './typeGuards';
+import { useEffect } from 'react';
+import { isSameCadModel, isSamePointCloudModel } from '../../utilities/isSameModel';
+
+export function useRemoveNonReferencedModels(
+  addOptions: TypedReveal3DModel[],
+  image360CollectionAddOptions: AddImageCollection360Options[],
+  viewer: Cognite3DViewer
+): void {
+  useEffect(() => {
+    const nonReferencedModels = findNonReferencedModels(addOptions, viewer);
+
+    nonReferencedModels.forEach((model) => {
+      viewer.removeModel(model);
+    });
+
+    const nonReferencedCollections = findNonReferencedCollections(
+      image360CollectionAddOptions,
+      viewer
+    );
+
+    nonReferencedCollections.forEach((collection) => {
+      viewer.remove360ImageSet(collection);
+    });
+  }, [addOptions]);
+}
+
+function findNonReferencedModels(
+  addOptions: TypedReveal3DModel[],
+  viewer: Cognite3DViewer
+): CogniteModel[] {
+  const models = viewer.models;
+  const addOptionsSet = new Set(addOptions);
+
+  return models.filter((model) => {
+    const correspondingAddOptions = [...addOptionsSet.values()].find((options) => {
```

Comment:
`[...addOptionsSet.values()]` array is created every loop, we can create this array outside the loop and use it

Reply:
No, because the set changes for every iteration. I can instead use a normal Array, and remove